Configuration file split proposal (#9236)
Mbed-TLS/mbedtls: docs/proposed/config-split.md

yet part of the PSA cryptography API (like LMS or PK).

It is followed by the "Data format support" section that contains configuration
options of utilities related to various data formats (like base64 or ASN1 APIs).
These utilities aim to facilitate the usage of the PSA cryptography API in other
cryptography projects.

Compared to Mbed TLS, the cryptography code in TF-PSA-Crypto is not located
in a single directory but split between the PSA core (core directory) and the
PSA builtin drivers (drivers/builtin/src directory). This is reflected in
tf_psa_crypto_config.h with two sections respectively named "PSA core" and
"Builtin drivers".

Finally, the last section named "Legacy cryptography" contains the configuration
options that will eventually be removed as duplicates of PSA_WANT_\* and
MBEDTLS_PSA_ACCEL_\* configuration options.

By contrast to mbedtls_config.h, tf_psa_crypto_config.h does not contain a
section like the "Module configuration options" one containing non boolean
configuration options. The configuration options that are not boolean are
located in the same section as the boolean option they are associated to.

> Open question: do we group them into a subsection?

Reviewer: > I wonder if the separation between boolean vs non-boolean was for the benefit of config.py

It dates from the PolarSSL days — so you tell me? `config.py` cares about sections because it's backward-compatible with `config.pl`. The non-boolean section is skipped by `config.py full`.

This is something we should probably change. More than once we've accidentally put a boolean option in the wrong section and it didn't get activated by `full`. It'll potentially affect the set of configurations that we're testing, but a new major release is a good time for that.

> Open question: do we group them into a subsection?

Reviewer: Uncommenting everything with `config.py` full would be a good thing. And on a related note [we mostly aren't testing non-boolean options](https://github.com/Mbed-TLS/mbedtls/issues/4048). From an architectural perspective, I'm all for making sections purely for the sake of documentation, and having them ignored by tooling. However, from a planning perspective, we may have to resolve CI complaints when we do this.

> Open question: do we group them into a subsection?

Author: Thanks for your comments. I won't have a lot of time to split the file (before the end of September) and address all the consequences of that. Thus I may need to be pragmatic and group the non boolean options into subsections, we will see. I think we have a consensus on keeping non boolean options in the same sections as their corresponding boolean options.

## Repartition of the configuration options

### In tf_psa_crypto_config.h, we have:
* SECTION "Platform abstraction layer"
#define MBEDTLS_HAVE_TIME
#define MBEDTLS_HAVE_TIME_DATE
//#define MBEDTLS_PLATFORM_MEMORY
//#define MBEDTLS_PLATFORM_NO_STD_FUNCTIONS
//#define MBEDTLS_PLATFORM_SETBUF_ALT
//#define MBEDTLS_PLATFORM_EXIT_ALT
//#define MBEDTLS_PLATFORM_TIME_ALT
//#define MBEDTLS_PLATFORM_FPRINTF_ALT
//#define MBEDTLS_PLATFORM_PRINTF_ALT
//#define MBEDTLS_PLATFORM_SNPRINTF_ALT
//#define MBEDTLS_PLATFORM_VSNPRINTF_ALT
//#define MBEDTLS_PLATFORM_NV_SEED_ALT
//#define MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT
//#define MBEDTLS_PLATFORM_MS_TIME_ALT
//#define MBEDTLS_PLATFORM_GMTIME_R_ALT
//#define MBEDTLS_PLATFORM_ZEROIZE_ALT

> #define MBEDTLS_FS_IO

Author: I've tried it and yes all together it looks better. I've kept the boolean and non boolean options separated though.

//#define MBEDTLS_MEMORY_DEBUG
//#define MBEDTLS_MEMORY_BACKTRACE
//#define MBEDTLS_THREADING_ALT
//#define MBEDTLS_THREADING_PTHREAD
#define MBEDTLS_PLATFORM_C
//#define MBEDTLS_THREADING_C

> #define MBEDTLS_TIMING_C

Reviewer 2: I think that belongs in TLS: the only public functions there are not generic at all but made specifically for `mbedtls_ssl_set_timer_cb()`. So at this point really the module should be renamed to `ssl_timers.c` and the options accordingly. The renaming is out of scope here (again, 4.0?) but I think where this option goes is in scope.

//#define MBEDTLS_PLATFORM_SNPRINTF_MACRO    snprintf
//#define MBEDTLS_PLATFORM_VSNPRINTF_MACRO    vsnprintf
//#define MBEDTLS_PLATFORM_NV_SEED_READ_MACRO   mbedtls_platform_std_nv_seed_read
//#define MBEDTLS_PLATFORM_NV_SEED_WRITE_MACRO  mbedtls_platform_std_nv_seed_write
//#define MBEDTLS_PLATFORM_MS_TIME_TYPE_MACRO   int64_t
//#define MBEDTLS_PRINTF_MS_TIME    PRId64
//#define MBEDTLS_MEMORY_ALIGN_MULTIPLE      4

* SECTION "General and test configuration options"
//#define MBEDTLS_PSA_CRYPTO_CONFIG_FILE "psa/crypto_config.h"
//#define MBEDTLS_PSA_CRYPTO_USER_CONFIG_FILE "/dev/null"
//#define MBEDTLS_DEPRECATED_WARNING
//#define MBEDTLS_DEPRECATED_REMOVED
//#define MBEDTLS_CHECK_RETURN_WARNING
#define MBEDTLS_SELF_TEST
//#define MBEDTLS_TEST_CONSTANT_FLOW_MEMSAN
//#define MBEDTLS_TEST_CONSTANT_FLOW_VALGRIND
//#define MBEDTLS_TEST_HOOKS

//#define MBEDTLS_CHECK_RETURN __attribute__((__warn_unused_result__))
//#define MBEDTLS_IGNORE_RETURN( result ) ((void) !(result))

* SECTION "Cryptographic mechanism selection (PSA API)"
PSA_WANT_\* macros as in current crypto_config.h.

* SECTION "Cryptographic mechanism selection (extended API)"

> #define MBEDTLS_CIPHER_C

Reviewer 2: I don't think `CIPHER_C` and the DRBG ones belong here, as those won't be public interfaces in 4.0 AFAIU.

Author: Moved to the "Builtin drivers" section.

//#define MBEDTLS_CTR_DRBG_USE_128_BIT_KEY
#define MBEDTLS_CTR_DRBG_C
#define MBEDTLS_HMAC_DRBG_C
#define MBEDTLS_LMS_C
//#define MBEDTLS_LMS_PRIVATE
#define MBEDTLS_MD_C
#define MBEDTLS_NIST_KW_C
#define MBEDTLS_PK_PARSE_EC_EXTENDED

> #define MBEDTLS_PK_PARSE_EC_COMPRESSED

Reviewer 2: Out of scope (naming): I wonder if we should keep those in the `MBEDTLS_PK` namespace - will they affect the new extended `psa_import()` API as well? If so, shouldn't they have a more neutral name?

> #define MBEDTLS_PK_RSA_ALT_SUPPORT

Reviewer 2: IIUC this is going to be removed in 4.0 as well.

Author: okay but I guess we should keep it for the time being and it is probably better to keep it here with other MBEDTLS_PK_ options.

#define MBEDTLS_PK_C
#define MBEDTLS_PK_PARSE_C
#define MBEDTLS_PK_WRITE_C

> #define MBEDTLS_PKCS5_C

Reviewer 2: Wasn't sure so I asked and it seems this and the next one will be internal as well (used by pkparse but not directly accessible to users) so I don't think their config options belong here.

Reviewer: PKCS5/PKCS12 _APIs_ will be internal. But the _options_ still make sense (maybe with a different name) as optional features (not everyone wants to pay for the code size if they don't need encrypted keys).

Reviewer 2: Ah, good point, missed that, thanks.

#define MBEDTLS_ECP_DP_SECP256K1_ENABLED
#define MBEDTLS_ECP_DP_BP256R1_ENABLED
#define MBEDTLS_ECP_DP_BP384R1_ENABLED
#define MBEDTLS_ECP_DP_BP512R1_ENABLED
#define MBEDTLS_ECP_DP_CURVE25519_ENABLED
#define MBEDTLS_ECP_DP_CURVE448_ENABLED
#define MBEDTLS_ECDSA_DETERMINISTIC
#define MBEDTLS_GENPRIME
#define MBEDTLS_PKCS1_V15
#define MBEDTLS_PKCS1_V21
//#define MBEDTLS_PSA_CRYPTO_CONFIG
#define MBEDTLS_AES_C
#define MBEDTLS_BIGNUM_C
#define MBEDTLS_CAMELLIA_C
#define MBEDTLS_ARIA_C
#define MBEDTLS_CCM_C
#define MBEDTLS_CHACHA20_C
#define MBEDTLS_CHACHAPOLY_C
#define MBEDTLS_CMAC_C
#define MBEDTLS_DES_C
#define MBEDTLS_DHM_C
#define MBEDTLS_ECDH_C
#define MBEDTLS_ECDSA_C
#define MBEDTLS_ECJPAKE_C
#define MBEDTLS_ECP_C
#define MBEDTLS_GCM_C
#define MBEDTLS_HKDF_C
#define MBEDTLS_MD5_C
#define MBEDTLS_PADLOCK_C
#define MBEDTLS_POLY1305_C
//#define MBEDTLS_PSA_CRYPTO_SE_C
#define MBEDTLS_RIPEMD160_C
#define MBEDTLS_RSA_C
#define MBEDTLS_SHA1_C
#define MBEDTLS_SHA224_C
#define MBEDTLS_SHA256_C
#define MBEDTLS_SHA384_C
#define MBEDTLS_SHA512_C
#define MBEDTLS_SHA3_C

### In mbedtls_config.h, we have:
* SECTION "Platform abstraction layer"
#define MBEDTLS_NET_C

> * SECTION "Mbed TLS feature support"

Reviewer 2: This is a bit beyond scope (perhaps more of a 4.0 item), but I'd really like to separate between X.509, TLS, and the rest (error, version).

> //#define MBEDTLS_CIPHER_NULL_CIPHER

Reviewer 2: Note: this should be made into a TLS option and renamed. It's not about "null cipher" in the no-longer-public, it's about integrity-only ciphersuites in TLS. (Not for the present PR of course, just pointing out as I see it, and let's create an issue to track it if we don't have one already.)

Reviewer: We already have several issues tracking renamings of options. Let's keep that out of scope here.

Reviewer 2: Yes, sorry, forgot to say this was out of scope and probably more for 4.0.

#define MBEDTLS_SSL_ALL_ALERT_MESSAGES
#define MBEDTLS_SSL_DTLS_CONNECTION_ID
#define MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT 0
//#define MBEDTLS_SSL_ASYNC_PRIVATE
#define MBEDTLS_SSL_CONTEXT_SERIALIZATION
//#define MBEDTLS_SSL_DEBUG_ALL
#define MBEDTLS_SSL_ENCRYPT_THEN_MAC
#define MBEDTLS_SSL_EXTENDED_MASTER_SECRET
#define MBEDTLS_SSL_KEEP_PEER_CERTIFICATE
#define MBEDTLS_SSL_RENEGOTIATION
#define MBEDTLS_SSL_MAX_FRAGMENT_LENGTH
//#define MBEDTLS_SSL_RECORD_SIZE_LIMIT
#define MBEDTLS_SSL_PROTO_TLS1_2
#define MBEDTLS_SSL_PROTO_TLS1_3
#define MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
//#define MBEDTLS_SSL_EARLY_DATA
#define MBEDTLS_SSL_PROTO_DTLS
#define MBEDTLS_SSL_ALPN
#define MBEDTLS_SSL_DTLS_ANTI_REPLAY
#define MBEDTLS_SSL_DTLS_HELLO_VERIFY
//#define MBEDTLS_SSL_DTLS_SRTP
#define MBEDTLS_SSL_DTLS_CLIENT_PORT_REUSE
#define MBEDTLS_SSL_SESSION_TICKETS
#define MBEDTLS_SSL_SERVER_NAME_INDICATION
//#define MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH
//#define MBEDTLS_USE_PSA_CRYPTO

> #define MBEDTLS_VERSION_C

Reviewer 2: Question: do we want to have something similar to `version.h` on the TF-PSA-Crypto side?

Author: Yes eventually, see https://github.com/Mbed-TLS/mbedtls/pull/9236#discussion_r1640203897.

#define MBEDTLS_VERSION_FEATURES
//#define MBEDTLS_X509_TRUSTED_CERTIFICATE_CALLBACK

> //#define MBEDTLS_X509_REMOVE_INFO

Reviewer 2: Note (again probably more a 4.0 item): this is a negative macro. We don't like negative macros. Change to positive?

#define MBEDTLS_X509_RSASSA_PSS_SUPPORT

> * SECTION "Mbed TLS modules"

Reviewer 2: Note, again probably more 4.0: I don't think the division between "feature support" and "modules" is user-friendly. We're putting implementation details in the user's face. Some of these options shouldn't even be public IMO (`SSL_TLS_C`, `X509_USE_C`, `X509_CREATE_C`). IIRC we discussed some of this last time we were all in Cambridge, and ISTR we wanted to simplify this and remove `_C` from macro names too.

#define MBEDTLS_DEBUG_C
#define MBEDTLS_ERROR_C
#define MBEDTLS_PKCS7_C
#define MBEDTLS_SSL_CACHE_C
#define MBEDTLS_SSL_COOKIE_C
#define MBEDTLS_SSL_TICKET_C
#define MBEDTLS_SSL_CLI_C
#define MBEDTLS_SSL_SRV_C
#define MBEDTLS_SSL_TLS_C
#define MBEDTLS_X509_USE_C
#define MBEDTLS_X509_CRT_PARSE_C
#define MBEDTLS_X509_CRL_PARSE_C
#define MBEDTLS_X509_CSR_PARSE_C
#define MBEDTLS_X509_CREATE_C
#define MBEDTLS_X509_CRT_WRITE_C
#define MBEDTLS_X509_CSR_WRITE_C

* SECTION "General configuration options"

> //#define MBEDTLS_CONFIG_FILE "mbedtls/mbedtls_config.h"

Reviewer 2: Seems like a weird thing to have in `mbedtls/mbedtls_config.h`.

Reviewer: How else would we advertize its existence?

Reviewer 2: Hmm, OK. Out of scope anyway.

Configuration file split
========================

> ## Why splitting the configuration file?

Reviewer: Minor:
```suggestion
## Why split the configuration file?
```

The objective of the repository split is to reach the point where in Mbed TLS

> all the cryptography code and its tests are located in a tf-psa-crypto

Reviewer: Minor:
```suggestion
all the cryptography code and its tests are located in a `tf-psa-crypto`
```

directory that just contains the TF-PSA-Crypto repository as a submodule.
The cryptography APIs exposed by Mbed TLS are just the TF-PSA-Crypto ones.
Mbed TLS relies solely on the TF-PSA-Crypto build system to build its
cryptography library and its tests.

The TF-PSA-Crypto configuration file `tf_psa_crypto_config.h` configures
entirely the cryptography interface exposed by Mbed TLS through TF-PSA-Crypto.

> Mbed TLS configuration is splitted in two files: `mbedtls_config.h` for TLS and

Reviewer: Minor:
```suggestion
Mbed TLS configuration is split in two files: `mbedtls_config.h` for TLS and
```

x509, `tf_psa_crypto_config.h` for the cryptography.

## How do we split the configuration file?

> We extend the so called PSA cryptographic configuration scheme based on

Reviewer: Minor:
```suggestion
We extend the so-called PSA cryptographic configuration scheme based on
```

`mbedtls_config.h` and `crypto_config.h`. The configuration file `crypto_config.h`
is extended to become the TF-PSA-Crypto configuration file, `mbedtls_config.h`
becomes the configuration file for the TLS and x509 libraries. All the options
to select the cryptographic mechanisms and to configure their implementation
are moved from `mbedtls_config.h` to `(tf_psa_)crypto_config.h`.

The configuration options that are relevant to both Mbed TLS and TF-PSA-Crypto
like platform or system ones are moved to `(tf_psa_)crypto_config.h`. That way
they are available in both repositories (as Mbed TLS includes
`tf_psa_crypto_config.h`) without duplication. Later, we may duplicate or
create aliases for some of them to align with the naming conventions of the
repositories.

> The layout of options into sections in `mbedtls_config.h` does not suit

Reviewer: It's not very clear what you mean: that `tf_psa_crypto_config.h` can't have the same layout as the 4.x `mbedtls_config.h` that you describe below, or that `tf_psa_crypto_config.h` can't have the same layout as the 3.x `mbedtls_config.h`? I read your sentence as the former, but in context it seems that you're arguing the latter (we can't keep the existing layout, hence we've designed a new layout).

Apart from sections that are intrinsically specific to one project (e.g. drivers in TF-PSA-Crypto, X.509 vs TLS in Mbed TLS), I don't see why the layouts would need to differ.

Author: I've changed that completely.

TF-PSA-Crypto well thus the configuration options `tf_psa_crypto_config.h` are
organized into different sections (see below).

## Configuration files and `config.py`

Each repository contains a `config.py` script to create and modify
configurations.

In Mbed TLS, `config.py` handles both `mbedtls_config.h` and
`tf_psa_crypto_config.h`. It can set or unset TLS, x509 and cryptographic
configuration options without having to specify the configuration file the
options belong to. Commands like full and baremetal affect both configuration
files.

In TF-PSA-Crypto, `config.py` addresses only `tf_psa_crypto_config.h`.

## Sections in `tf_psa_crypto_config.h`

The `tf_psa_crypto_config.h` configuration file is organized into eight
sections.

The pre-split `mbedtls_config.h` configuration file contains configuration
options that apply to the whole code base (TLS, x509, crypto and tests) mostly
related to the platform abstraction layer and testing. In
`tf_psa_crypto_config.h` these configurations options are organized into two
sections, one for the platform abstraction layer options and one for the others,

> respectively named "Platform abstraction layer" and

Reviewer: Minor: it would be nice to have links to the sections, e.g.
```suggestion
respectively named ["Platform abstraction layer"](#section-platform-abstraction-layer) and
```

"General and test configuration options".

Then, the "Cryptographic mechanism selection (PSA API)" section is the
equivalent of the pre-split `crypto_config.h` configuration file containing the
PSA_WANT_ prefixed macros.

The following section named "Cryptographic mechanism selection (extended API)"
contains the configuration options for the cryptography mechanisms that are not
yet part of the PSA cryptography API (like LMS or PK).

It is followed by the "Data format support" section that contains configuration

> options of utilities related to various data formats (like base64 or ASN1 APIs).

Reviewer: Minor:
```suggestion
options of utilities related to various data formats (like Base64 or ASN.1 APIs).
```

These utilities aim to facilitate the usage of the PSA cryptography API in other
cryptography projects.

Compared to Mbed TLS, the cryptography code in TF-PSA-Crypto is not located
in a single directory but split between the PSA core (core directory) and the
PSA builtin drivers (drivers/builtin/src directory). This is reflected in
`tf_psa_crypto_config.h` with two sections respectively named "PSA core" and
"Builtin drivers".

Finally, the last section named "Legacy cryptography" contains the configuration
options that will eventually be removed as duplicates of PSA_WANT_\* and
MBEDTLS_PSA_ACCEL_\* configuration options.

By contrast to `mbedtls_config.h`, `tf_psa_crypto_config.h` does not contain a

> section like the "Module configuration options" one containing non boolean

Reviewer: Why would `mbedtls_config.h` keep a separate section for "Module configuration options"? Either it's a problem with our tooling and both projects need to keep non-boolean options in separate sections, or the tooling is fine and I don't see why we wouldn't merge "Module configuration options" with "Mbed TLS feature support".

Reviewer: After discussion and experiment, it _is_ a problem with our tooling, but one that it's easier to solve than work around.

https://github.com/Mbed-TLS/mbedtls/pull/9604 removes the reliance of `config.py` on which section a setting is in.

Author: No section dedicated to non boolean option anymore, both in tf_psa_crypto_config.h and mbedtls_config.h.

configuration options. The configuration options that are not boolean are
located in the same section as the boolean option they are associated to.

> Open question: do we group them into a subsection?

Reviewer: Let's not add a concept of subsections. I can't see a reason for an extra level.

//#define MBEDTLS_PLATFORM_STD_CALLOC        calloc
//#define MBEDTLS_PLATFORM_STD_EXIT            exit
//#define MBEDTLS_PLATFORM_STD_EXIT_FAILURE       1
//#define MBEDTLS_PLATFORM_STD_EXIT_SUCCESS       0
//#define MBEDTLS_PLATFORM_STD_FPRINTF      fprintf
//#define MBEDTLS_PLATFORM_STD_FREE            free
//#define MBEDTLS_PLATFORM_STD_MEM_HDR   <stdlib.h>
//#define MBEDTLS_PLATFORM_STD_NV_SEED_FILE  "seedfile"
//#define MBEDTLS_PLATFORM_STD_NV_SEED_READ   mbedtls_platform_std_nv_seed_read
//#define MBEDTLS_PLATFORM_STD_NV_SEED_WRITE  mbedtls_platform_std_nv_seed_write
//#define MBEDTLS_PLATFORM_STD_PRINTF        printf
//#define MBEDTLS_PLATFORM_STD_SETBUF      setbuf
//#define MBEDTLS_PLATFORM_STD_SNPRINTF    snprintf
//#define MBEDTLS_PLATFORM_STD_TIME            time
//#define MBEDTLS_PLATFORM_SNPRINTF_MACRO    snprintf
//#define MBEDTLS_PLATFORM_TIME_MACRO            time
//#define MBEDTLS_PLATFORM_TIME_TYPE_MACRO       time_t
//#define MBEDTLS_PLATFORM_VSNPRINTF_MACRO    vsnprintf
```

#### SECTION "General and test configuration options"
```
//#define MBEDTLS_CHECK_RETURN_WARNING
//#define MBEDTLS_DEPRECATED_REMOVED
//#define MBEDTLS_DEPRECATED_WARNING

> //#define MBEDTLS_PSA_CRYPTO_CONFIG_FILE "psa/crypto_config.h"

Reviewer: Following up on https://github.com/Mbed-TLS/mbedtls/pull/9236#discussion_r1640202307:

> > Presumably MBEDTLS_PSA_CRYPTO_CONFIG_FILE will become TF_PSA_CRYPTO_CONFIG_FILE and MBEDTLS_PSA_CRYPTO_USER_CONFIG_FILE will become TF_PSA_CRYPTO_USER_CONFIG_FILE. (…)
>
> I haven't renamed them yet. Not sure when we will do it. It seems to me it can be done even after the repo split.

Agreed, but please add a note to this document, because when we look at this document in isolation, it seems inconsistent to remove `psa/crypto_config.h` as a file but keep an option that refers to it (both in its name and its default value).

In other words, I think renaming these options is part of the changes specified by this architectural document. But in terms of planning, it's a distinct task, and it can be done after the files have moved to their respective repositories.

Author: I've renamed them eventually and added a note as these are the only options to be renamed. And indeed yes, when we do it that's another question.

#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
#define MBEDTLS_SSL_SERVER_NAME_INDICATION
#define MBEDTLS_SSL_SESSION_TICKETS
//#define MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH
//#define MBEDTLS_USE_PSA_CRYPTO
#define MBEDTLS_VERSION_C
#define MBEDTLS_VERSION_FEATURES
//#define MBEDTLS_X509_TRUSTED_CERTIFICATE_CALLBACK
//#define MBEDTLS_X509_REMOVE_INFO
#define MBEDTLS_X509_RSASSA_PSS_SUPPORT
```

#### SECTION "Mbed TLS modules"
```
#define MBEDTLS_DEBUG_C
#define MBEDTLS_ERROR_C
#define MBEDTLS_PKCS7_C
#define MBEDTLS_SSL_CACHE_C
#define MBEDTLS_SSL_COOKIE_C
#define MBEDTLS_SSL_CLI_C
#define MBEDTLS_SSL_SRV_C
#define MBEDTLS_SSL_TICKET_C
#define MBEDTLS_SSL_TLS_C
#define MBEDTLS_X509_CREATE_C
#define MBEDTLS_X509_CRL_PARSE_C
#define MBEDTLS_X509_CRT_PARSE_C
#define MBEDTLS_X509_CRT_WRITE_C
#define MBEDTLS_X509_CSR_PARSE_C
#define MBEDTLS_X509_CSR_WRITE_C
#define MBEDTLS_X509_USE_C
```

#### SECTION "General configuration options"
```
//#define MBEDTLS_CONFIG_FILE "mbedtls/mbedtls_config.h"
//#define MBEDTLS_USER_CONFIG_FILE "/dev/null"
```

> #### SECTION "Module configuration options"

Reviewer: Following the structure of the crypto config file, I think we should change the grouping of `mbedtls_config.h`:
"Mbed TLS feature support"
 + "Mbed TLS modules" + "Module configuration options" → "X.509 feature selection" + "TLS feature selection"

> Configuration file split

Reviewer: Please add this file to `docs/proposed/Makefile`.

(Unfortunately pandoc isn't 100% compatible with GitHub: it has slightly different rules for section anchor names, even with `-f gfm`, at least as of version 2.9.2.1. But it's close, including most section anchors, tables, etc.)

section that contains configuration options of utilities related to various data
formats (like Base64 or ASN.1 APIs). These utilities aim to facilitate the
usage of the PSA cryptography API in other cryptography projects.

Compared to Mbed TLS, the cryptography code in TF-PSA-Crypto is not located
in a single directory but split between the PSA core (core directory) and the
PSA builtin drivers (drivers/builtin/src directory). This is reflected in
`tf_psa_crypto_config.h` with two sections respectively named ["PSA core"](#section-psa-core)
and ["Builtin drivers"](#section-builtin-drivers).

Finally, the last section named ["Legacy cryptography"](#section-legacy-cryptography)
contains the configuration options that will eventually be removed as duplicates
of PSA_WANT_\* and MBEDTLS_PSA_ACCEL_\* configuration options.

## Sections in `mbedtls_config.h`

The sections in `mbedtls_config.h` are reorganized to be better aligned with
the ones in `tf_psa_crypto_config.h`. The main change is the reorganization
of the "Mbed TLS modules", "Mbed TLS feature support" and
"Module configuration options" sections into the
["TLS feature selection"](#section-tls-feature-selection) and

> ["X.509 feature selection"](#section-x.509-feature-selection) sections. That

Reviewer: ```suggestion
["X.509 feature selection"](#section-x-509-feature-selection) sections. That
```

//#define MBEDTLS_PLATFORM_STD_CALLOC        calloc
//#define MBEDTLS_PLATFORM_STD_EXIT            exit
//#define MBEDTLS_PLATFORM_STD_EXIT_FAILURE       1
//#define MBEDTLS_PLATFORM_STD_EXIT_SUCCESS       0
//#define MBEDTLS_PLATFORM_STD_FPRINTF      fprintf
//#define MBEDTLS_PLATFORM_STD_FREE            free
//#define MBEDTLS_PLATFORM_STD_MEM_HDR   <stdlib.h>
//#define MBEDTLS_PLATFORM_STD_NV_SEED_FILE  "seedfile"
//#define MBEDTLS_PLATFORM_STD_NV_SEED_READ   mbedtls_platform_std_nv_seed_read
//#define MBEDTLS_PLATFORM_STD_NV_SEED_WRITE  mbedtls_platform_std_nv_seed_write
//#define MBEDTLS_PLATFORM_STD_PRINTF        printf
//#define MBEDTLS_PLATFORM_STD_SETBUF      setbuf
//#define MBEDTLS_PLATFORM_STD_SNPRINTF    snprintf
//#define MBEDTLS_PLATFORM_STD_TIME            time
//#define MBEDTLS_PLATFORM_TIME_MACRO            time
//#define MBEDTLS_PLATFORM_TIME_TYPE_MACRO       time_t
//#define MBEDTLS_PLATFORM_VSNPRINTF_MACRO    vsnprintf
//#define MBEDTLS_PRINTF_MS_TIME    PRId64
```

#### SECTION General and test configuration options
Note: for consistency with the configuration file name change from
`crypto_config.h` to `tf_psa_crypto_config.h`, the configuration options

> MBEDTLS_PSA_CRYPTO_CONFIG_FILE and MBEDTLS_PSA_CRYPTO_USER_CONFIG_FILE are
> respectively renamed TF_PSA_CRYPTO_CONFIG_FILE and TF_PSA_CRYPTO_USER_CONFIG_FILE.

Reviewer: Minor:
```suggestion
`MBEDTLS_PSA_CRYPTO_CONFIG_FILE` and `MBEDTLS_PSA_CRYPTO_USER_CONFIG_FILE` are
respectively renamed `TF_PSA_CRYPTO_CONFIG_FILE` and `TF_PSA_CRYPTO_USER_CONFIG_FILE`.
```

#define MBEDTLS_RSA_C  
#define MBEDTLS_SHA1_C  
#define MBEDTLS_SHA224_C  
#define MBEDTLS_SHA256_C  
#define MBEDTLS_SHA384_C  
#define MBEDTLS_SHA512_C  
#define MBEDTLS_SHA3_C  

### In tf_psa_crypto_config.h, we have:  
* SECTION "Platform abstraction layer configuration options"  
#define MBEDTLS_HAVE_TIME  
#define MBEDTLS_HAVE_TIME_DATE  
//#define MBEDTLS_PLATFORM_MEMORY  
//#define MBEDTLS_PLATFORM_NO_STD_FUNCTIONS  
//#define MBEDTLS_PLATFORM_SETBUF_ALT  
//#define MBEDTLS_PLATFORM_EXIT_ALT  
//#define MBEDTLS_PLATFORM_TIME_ALT  
//#define MBEDTLS_PLATFORM_FPRINTF_ALT  
//#define MBEDTLS_PLATFORM_PRINTF_ALT  
//#define MBEDTLS_PLATFORM_SNPRINTF_ALT  
//#define MBEDTLS_PLATFORM_VSNPRINTF_ALT  
//#define MBEDTLS_PLATFORM_NV_SEED_ALT  
//#define MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT  
//#define MBEDTLS_PLATFORM_MS_TIME_ALT  
//#define MBEDTLS_PLATFORM_GMTIME_R_ALT  
//#define MBEDTLS_PLATFORM_ZEROIZE_ALT  
#define MBEDTLS_FS_IO  
//#define MBEDTLS_MEMORY_DEBUG  
//#define MBEDTLS_MEMORY_BACKTRACE  
//#define MBEDTLS_THREADING_ALT ???  
//#define MBEDTLS_THREADING_PTHREAD  
#define MBEDTLS_PLATFORM_C  
//#define MBEDTLS_THREADING_C  
#define MBEDTLS_TIMING_C  

> //#define MBEDTLS_TIMING_ALT ???  

Author: I've kept it here for the time being because indeed of would we do without it in the benchmark program?

//#define MBEDTLS_PLATFORM_STD_NV_SEED_FILE  "seedfile"  
//#define MBEDTLS_PLATFORM_CALLOC_MACRO        calloc  
//#define MBEDTLS_PLATFORM_FREE_MACRO            free  
//#define MBEDTLS_PLATFORM_EXIT_MACRO            exit  
//#define MBEDTLS_PLATFORM_SETBUF_MACRO      setbuf  
//#define MBEDTLS_PLATFORM_TIME_MACRO            time  
//#define MBEDTLS_PLATFORM_TIME_TYPE_MACRO       time_t  
//#define MBEDTLS_PLATFORM_FPRINTF_MACRO      fprintf  
//#define MBEDTLS_PLATFORM_PRINTF_MACRO        printf  
//#define MBEDTLS_PLATFORM_SNPRINTF_MACRO    snprintf  
//#define MBEDTLS_PLATFORM_VSNPRINTF_MACRO    vsnprintf  
//#define MBEDTLS_PLATFORM_NV_SEED_READ_MACRO   mbedtls_platform_std_nv_seed_read  
//#define MBEDTLS_PLATFORM_NV_SEED_WRITE_MACRO  mbedtls_platform_std_nv_seed_write  
//#define MBEDTLS_PLATFORM_MS_TIME_TYPE_MACRO   int64_t  
//#define MBEDTLS_PRINTF_MS_TIME    PRId64  
//#define MBEDTLS_MEMORY_ALIGN_MULTIPLE      4  
//#define MBEDTLS_CHECK_RETURN __attribute__((__warn_unused_result__))  
//#define MBEDTLS_IGNORE_RETURN( result ) ((void) !(result))  

* SECTION "General and test configuration options"  

> //#define MBEDTLS_PSA_CRYPTO_CONFIG_FILE "psa/crypto_config.h"  

Author: I haven't renamed them yet. Not sure when we will do it. It seems to me it can be done even after the repo split.

//#define MBEDTLS_PSA_CRYPTO_USER_CONFIG_FILE "/dev/null"  
//#define MBEDTLS_DEPRECATED_WARNING  
//#define MBEDTLS_DEPRECATED_REMOVED  
//#define MBEDTLS_CHECK_RETURN_WARNING  
//#define MBEDTLS_TEST_CONSTANT_FLOW_MEMSAN  
//#define MBEDTLS_TEST_CONSTANT_FLOW_VALGRIND  

> //#define MBEDTLS_TEST_HOOKS  

Author: I have kept them in the "General and test configuration options" section that is quite small. That seems acceptable especially if we think they should be removed eventually.

#define MBEDTLS_VERSION_C  

> #define MBEDTLS_VERSION_FEATURES  

Author: I've restored them in mbedtls_config.h and added TF_PSA_CRYPTO_ERROR/VERSION/VERSION_FEATURES configuration options in tf_psa_crypto_config.h.

* SECTION "PSA cryptography API configuration options"  

> include/psa/crypto_config.h  

Author: > So are we keeping `psa/crypto_config.h`? That contradicts the statements above that list `tf_psa_crypto_config.h` as the sole configuration file for TF-PSA-Crypto.

That's not what I wanted to express here. This was just meaning: all the content of crypto_config.h.

//#define MBEDTLS_AES_ROM_TABLES  
//#define MBEDTLS_AES_FEWER_TABLES  
//#define MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH  
//#define MBEDTLS_AES_USE_HARDWARE_ONLY  
//#define MBEDTLS_CAMELLIA_SMALL_MEMORY  
//#define MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED  
#define MBEDTLS_ECP_NIST_OPTIM  
//#define MBEDTLS_ECP_RESTARTABLE ???  
//#define MBEDTLS_ECP_WITH_MPI_UINT  
//#define MBEDTLS_PSA_P256M_DRIVER_ENABLED  
//#define MBEDTLS_SHA256_SMALLER  
//#define MBEDTLS_SHA512_SMALLER  
//#define MBEDTLS_RSA_NO_CRT  
#define MBEDTLS_SELF_TEST  
//#define MBEDTLS_BLOCK_CIPHER_NO_DECRYPT  
//#define MBEDTLS_GCM_LARGE_TABLE  
//#define MBEDTLS_SHA256_USE_ARMV8_A_CRYPTO_IF_PRESENT  
//#define MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT  
//#define MBEDTLS_SHA256_USE_ARMV8_A_CRYPTO_ONLY  
//#define MBEDTLS_SHA256_USE_A64_CRYPTO_ONLY  
//#define MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT  
//#define MBEDTLS_SHA512_USE_A64_CRYPTO_ONLY  
//#define MBEDTLS_MPI_WINDOW_SIZE            2  
//#define MBEDTLS_MPI_MAX_SIZE            1024  
//#define MBEDTLS_ECP_WINDOW_SIZE            4  
//#define MBEDTLS_ECP_FIXED_POINT_OPTIM      1  
//#define MBEDTLS_RSA_GEN_KEY_MIN_BITS            1024  

> * SECTION "Beyond the current PSA cryptography API configuration options."  

Author: I prefer to keep a section dedicated to the PSA_WANT_* macros thus still two sections.

#define MBEDTLS_SSL_CONTEXT_SERIALIZATION  
//#define MBEDTLS_SSL_DEBUG_ALL  
#define MBEDTLS_SSL_ENCRYPT_THEN_MAC  
#define MBEDTLS_SSL_EXTENDED_MASTER_SECRET  
#define MBEDTLS_SSL_KEEP_PEER_CERTIFICATE  
#define MBEDTLS_SSL_RENEGOTIATION  
#define MBEDTLS_SSL_MAX_FRAGMENT_LENGTH  
//#define MBEDTLS_SSL_RECORD_SIZE_LIMIT  
#define MBEDTLS_SSL_PROTO_TLS1_2  
#define MBEDTLS_SSL_PROTO_TLS1_3  
#define MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE  
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED  
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED  
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED  
//#define MBEDTLS_SSL_EARLY_DATA  
#define MBEDTLS_SSL_PROTO_DTLS  
#define MBEDTLS_SSL_ALPN  
#define MBEDTLS_SSL_DTLS_ANTI_REPLAY  
#define MBEDTLS_SSL_DTLS_HELLO_VERIFY  
//#define MBEDTLS_SSL_DTLS_SRTP  
#define MBEDTLS_SSL_DTLS_CLIENT_PORT_REUSE  
#define MBEDTLS_SSL_SESSION_TICKETS  
#define MBEDTLS_SSL_SERVER_NAME_INDICATION  
//#define MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH  
//#define MBEDTLS_USE_PSA_CRYPTO  
//#define MBEDTLS_X509_TRUSTED_CERTIFICATE_CALLBACK  
//#define MBEDTLS_X509_REMOVE_INFO  
#define MBEDTLS_X509_RSASSA_PSS_SUPPORT  

> * SECTION "Mbed TLS modules"  

Author: I guess nothing to change for the comment. I've just kept this section in mbedtls_config.h.

#define MBEDTLS_X509_CSR_WRITE_C  

* SECTION "General configuration options"  
//#define MBEDTLS_CONFIG_FILE "mbedtls/mbedtls_config.h"  
//#define MBEDTLS_USER_CONFIG_FILE "/dev/null"  

* SECTION "Module configuration options"  
//#define MBEDTLS_SSL_CACHE_DEFAULT_TIMEOUT       86400  
//#define MBEDTLS_SSL_CACHE_DEFAULT_MAX_ENTRIES      50  
//#define MBEDTLS_SSL_IN_CONTENT_LEN              16384  
//#define MBEDTLS_SSL_CID_IN_LEN_MAX 32  
//#define MBEDTLS_SSL_CID_OUT_LEN_MAX 32  
//#define MBEDTLS_SSL_CID_TLS1_3_PADDING_GRANULARITY 16  
//#define MBEDTLS_SSL_OUT_CONTENT_LEN             16384  
//#define MBEDTLS_SSL_DTLS_MAX_BUFFERING             32768  
//#define MBEDTLS_PSK_MAX_LEN               32  
//#define MBEDTLS_SSL_COOKIE_TIMEOUT        60  
//#define MBEDTLS_SSL_CIPHERSUITES MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256  
//#define MBEDTLS_SSL_MAX_EARLY_DATA_SIZE        1024  
//#define MBEDTLS_SSL_TLS1_3_TICKET_AGE_TOLERANCE 6000  
//#define MBEDTLS_SSL_TLS1_3_TICKET_NONCE_LENGTH 32  
//#define MBEDTLS_SSL_TLS1_3_DEFAULT_NEW_SESSION_TICKETS 1  
//#define MBEDTLS_X509_MAX_INTERMEDIATE_CA   8  

> //#define MBEDTLS_X509_MAX_FILE_PATH_LEN     512  

Reviewer: Please add a section on `config.py` as a configuration interface. Which file(s) will it edit?

//#define MBEDTLS_CTR_DRBG_RESEED_INTERVAL        10000
//#define MBEDTLS_HMAC_DRBG_MAX_INPUT           256
//#define MBEDTLS_HMAC_DRBG_MAX_REQUEST        1024
//#define MBEDTLS_HMAC_DRBG_MAX_SEED_INPUT      384
//#define MBEDTLS_HMAC_DRBG_RESEED_INTERVAL   10000
//#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA256
```

#### SECTION Data format support
```
#define MBEDTLS_ASN1_PARSE_C
#define MBEDTLS_ASN1_WRITE_C
#define MBEDTLS_BASE64_C
#define MBEDTLS_OID_C
#define MBEDTLS_PEM_PARSE_C
#define MBEDTLS_PEM_WRITE_C
```

#### SECTION PSA core
```
#define MBEDTLS_ENTROPY_C
//#define MBEDTLS_ENTROPY_FORCE_SHA256
//#define MBEDTLS_ENTROPY_HARDWARE_ALT
//#define MBEDTLS_ENTROPY_NV_SEED
//#define MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES
//#define MBEDTLS_NO_PLATFORM_ENTROPY
//#define MBEDTLS_PSA_ASSUME_EXCLUSIVE_BUFFERS
//#define MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS
#define MBEDTLS_PSA_CRYPTO_C
//#define MBEDTLS_PSA_CRYPTO_CLIENT
//#define MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
//#define MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER
//#define MBEDTLS_PSA_CRYPTO_SPM
#define MBEDTLS_PSA_CRYPTO_STORAGE_C
//#define MBEDTLS_PSA_INJECT_ENTROPY
#define MBEDTLS_PSA_ITS_FILE_C

> //#define MBEDTLS_ENTROPY_MAX_GATHER                128

Reviewer 2: I'm not sure I understand why the DRBG options, including `MBEDTLS_PSA_HMAC_DRBG_MD_TYPE`, are in "Cryptographic mechanism selection (extended API)" but the `MBEDTLS_ENTROPY` options are here.

Not a blocker for the split as moving something internally between sections of the same file can be done later.

Actually most of the entropy/drbg options might actually move to "Built-in drivers" below, once we have completed driver support for the RNG subsystem.

Author: @gilles-peskine-arm posted a comment about that: https://github.com/Mbed-TLS/mbedtls/pull/9236#discussion_r1640211393.

Reviewer 2: Ah, I see. Thanks for the pointer.

#define MBEDTLS_ECP_DP_SECP224R1_ENABLED
#define MBEDTLS_ECP_DP_SECP256K1_ENABLED
#define MBEDTLS_ECP_DP_SECP256R1_ENABLED
#define MBEDTLS_ECP_DP_SECP384R1_ENABLED
#define MBEDTLS_ECP_DP_SECP521R1_ENABLED
#define MBEDTLS_ECDSA_C
#define MBEDTLS_ECDSA_DETERMINISTIC
#define MBEDTLS_ECJPAKE_C
#define MBEDTLS_GCM_C
#define MBEDTLS_GENPRIME
#define MBEDTLS_HKDF_C
#define MBEDTLS_HMAC_DRBG_C
#define MBEDTLS_MD5_C
#define MBEDTLS_PADLOCK_C
#define MBEDTLS_PKCS1_V15
#define MBEDTLS_PKCS1_V21
#define MBEDTLS_POLY1305_C
//#define MBEDTLS_PSA_CRYPTO_CONFIG
//#define MBEDTLS_PSA_CRYPTO_SE_C
#define MBEDTLS_RIPEMD160_C
#define MBEDTLS_RSA_C
#define MBEDTLS_SHA1_C
#define MBEDTLS_SHA224_C
#define MBEDTLS_SHA256_C
#define MBEDTLS_SHA384_C
#define MBEDTLS_SHA3_C
#define MBEDTLS_SHA512_C
```

### In `mbedtls_config.h`, we have:
#### SECTION Platform abstraction layer
```
#define MBEDTLS_NET_C
//#define MBEDTLS_TIMING_ALT
#define MBEDTLS_TIMING_C
```

#### SECTION General configuration options
```
//#define MBEDTLS_CONFIG_FILE "mbedtls/mbedtls_config.h"
//#define MBEDTLS_USER_CONFIG_FILE "/dev/null"
```

#### SECTION TLS feature selection
```
#define MBEDTLS_DEBUG_C

> #define MBEDTLS_ERROR_C

Reviewer 2: I don't think `ERROR_C` is specific to TLS, IMO it belongs in "General configuration options" above. (As opposed to `DEBUG_C` which really should be called `SSL_DEBUG_C`, see https://github.com/Mbed-TLS/mbedtls/issues/6859.)

Not a blocker for the split as moving something internally between sections of the same file can be done later.

#define MBEDTLS_SSL_CACHE_C
#define MBEDTLS_SSL_CLI_C
#define MBEDTLS_SSL_COOKIE_C
#define MBEDTLS_SSL_SRV_C
#define MBEDTLS_SSL_TICKET_C
#define MBEDTLS_SSL_TLS_C

> //#define MBEDTLS_PSK_MAX_LEN               32

Reviewer 2: Out of scope, but perhaps for 4.0 (ping @gilles-peskine-arm ) - should be renamed to `MBEDTLS_SSL_PSK_MAX_LEN` for clarity & consistency.

//#define MBEDTLS_SSL_CACHE_DEFAULT_TIMEOUT       86400
//#define MBEDTLS_SSL_CID_IN_LEN_MAX 32
//#define MBEDTLS_SSL_CID_OUT_LEN_MAX 32
//#define MBEDTLS_SSL_CID_TLS1_3_PADDING_GRANULARITY 16
//#define MBEDTLS_SSL_CIPHERSUITES MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
//#define MBEDTLS_SSL_COOKIE_TIMEOUT        60
//#define MBEDTLS_SSL_DTLS_MAX_BUFFERING             32768
//#define MBEDTLS_SSL_IN_CONTENT_LEN              16384
//#define MBEDTLS_SSL_MAX_EARLY_DATA_SIZE        1024
//#define MBEDTLS_SSL_OUT_CONTENT_LEN             16384
//#define MBEDTLS_SSL_TLS1_3_DEFAULT_NEW_SESSION_TICKETS 1
//#define MBEDTLS_SSL_TLS1_3_TICKET_AGE_TOLERANCE 6000
//#define MBEDTLS_SSL_TLS1_3_TICKET_NONCE_LENGTH 32
```

#### SECTION X.509 feature selection
```
#define MBEDTLS_PKCS7_C
#define MBEDTLS_X509_CREATE_C
#define MBEDTLS_X509_CRL_PARSE_C
#define MBEDTLS_X509_CRT_PARSE_C
#define MBEDTLS_X509_CRT_WRITE_C
#define MBEDTLS_X509_CSR_PARSE_C
#define MBEDTLS_X509_CSR_WRITE_C
#define MBEDTLS_X509_USE_C

//#define MBEDTLS_X509_MAX_FILE_PATH_LEN     512
//#define MBEDTLS_X509_MAX_INTERMEDIATE_CA   8
```

> #### SECTION Mbed TLS feature support

Reviewer 2: I don't understand this section. IMO everything here should be moved to one of the three sections above.

Looks like the sections above only have boolean options ending in `_C` and non-boolean option, and this section got the rests, ie boolean options not ending in `_C`. As I've said before, I think having some options named with `_C` and others not is really throwing implementation details* into the user's face, which is bad design. Separating options into sections based on this is IMO just as bad.

*Actually not even implementation details, historical details. Currently TLS files are organized as `ssl_version_side.c` mostly but for some reason "side" is a `_C` option (`CLI_C` vs `SRV_C`) while "version" is not (`PROTO_TLS1_x`). Regardless, we're exposing users to things that they shouldn't need to know.

As a user, "I want support for TLS clients" is extremely similar to "I want support for TLS 1.3" and there is no reason for the two of them to live in different categories.

Specifically, I think:
- `MBEDTLS_ERROR_STRERROR_DUMMY`, `MBEDTLS_VERSION_xxx` should go in "General configuration options"
- the `MBEDTLS_X509` option should go to X.509
- everything else to TLS.

Not a blocker for the split as moving something internally between sections of the same file can be done later.

Author: Yes I thought about merging "Mbed TLS feature support" while working on the commit Re-organize "Mbed TLS modules" and "Module configuration options . I am happy to reorganize further `mbedtls_config.h` post-split. @gilles-peskine-arm are you okay with that?

#define MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED
#define MBEDTLS_SSL_ALL_ALERT_MESSAGES
#define MBEDTLS_SSL_ALPN
//#define MBEDTLS_SSL_ASYNC_PRIVATE
#define MBEDTLS_SSL_CONTEXT_SERIALIZATION
//#define MBEDTLS_SSL_DEBUG_ALL
#define MBEDTLS_SSL_DTLS_ANTI_REPLAY
#define MBEDTLS_SSL_DTLS_CLIENT_PORT_REUSE
#define MBEDTLS_SSL_DTLS_CONNECTION_ID
#define MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT 0
#define MBEDTLS_SSL_DTLS_HELLO_VERIFY
//#define MBEDTLS_SSL_DTLS_SRTP
//#define MBEDTLS_SSL_EARLY_DATA
#define MBEDTLS_SSL_ENCRYPT_THEN_MAC
#define MBEDTLS_SSL_EXTENDED_MASTER_SECRET
#define MBEDTLS_SSL_KEEP_PEER_CERTIFICATE
#define MBEDTLS_SSL_MAX_FRAGMENT_LENGTH
#define MBEDTLS_SSL_PROTO_DTLS
#define MBEDTLS_SSL_PROTO_TLS1_2
#define MBEDTLS_SSL_PROTO_TLS1_3
//#define MBEDTLS_SSL_RECORD_SIZE_LIMIT
#define MBEDTLS_SSL_RENEGOTIATION
#define MBEDTLS_SSL_SERVER_NAME_INDICATION
#define MBEDTLS_SSL_SESSION_TICKETS
#define MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED
#define MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
//#define MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH

> //#define MBEDTLS_USE_PSA_CRYPTO

Reviewer 2: This influences PK, which is on the crypto side, so should it go in the crypto config file?

## Repartition of the configuration options

### In `tf_psa_crypto_config.h`, we have:

> * SECTION "Platform abstraction layer"

Reviewer 2: This will render to something unreadable: https://github.com/ronald-cron-arm/mbedtls/blob/config-file-split/docs/proposed/config-split.md#in-tf_psa_crypto_configh-we-have Suggestion:

    #### SECTION ...
    ```
    # define ...
    ```

Author: indeed, thanks.
